DOCS-2536: Refactor data service pages (#3473)
viamrobotics/docs: docs/services/data/capture-sync.md

---
title: "Data Capture and Sync"
linkTitle: "Data Capture and Sync"

> description: "Configure the data manager to capture data from your components and services and sync it to the cloud."

Reviewer: ```suggestion
description: "Configure the data management service to capture data from your components and services and sync it to the cloud."
```
If that's too long get rid of "from your components and services"

weight: 10
type: "docs"
tags: ["data management", "cloud", "sync", "capture"]
icon: true
images: ["/services/icons/data-capture.svg"]
no_list: true
aliases:
  - /services/data/capture/
  - /data/capture/
  - /build/micro-rdk/data_management/
  - /services/data/capture/
  - /services/data/cloud-sync/
  - /data/cloud-sync/
no_service: true
---

The data management service captures data from one or more {{< glossary_tooltip term_id="resource" text="resources" >}} locally, and syncs it to cloud storage when a connection to the cloud is available.
You can configure which data you want to capture, as well as the capture rate and the sync frequency.

> Data capture is frequently used with cloud sync.
> However, if you want to manage your machine's captured data yourself, you can enable only data capture without cloud sync.

Reviewer: Let's remove that here and mention it as a side note under data capture details
```suggestion
```

Get started with a quickstart guide or keep reading for more details.

{{< cards >}}
{{< card link="/get-started/collect-data/" class="green">}}
{{< /cards >}}

## Data capture details

> 

Reviewer: I think both this and the cloud sync details heading should go - we don't really want to separate data capture and sync mentally.

{{< tabs >}}
{{% tab name="viam-server" %}}

> 

Author: Why would this be on the viam-server tab specifically? Is this part of the assumption that all micro-viam-server users will use sync?

> 

Reviewer: I think you asked the data team about this right? If you capture data using viam-micro-server, you have captured it but you cannot access it. If you disconnect the device, it's gone. There's no way of getting at the data except to use cloud sync. So there's no point of using data capture without cloud sync

> 

Author: I see what you're saying. However, I'm not sure about the disconnecting part?? If it's in flash memory, isn't still there?

> 

Author: I am now confused because the internet says esp32 flash memory persists when power is cycled, but we have the line here in the docs to the contrary. Do you know for sure? Otherwise I'll ask about that.

> 

Reviewer: Hmm - I'm fairly certain it was SME-reviewed and that's what they said but what you say makes sense in that it should still be there. Maybe upon restarting `viam-micro` wipes flash storage or something.

I just re-read the convo you had with Gautham in the data management channel. He says "you won't have any means to retrieve it". By all means confirm it but if the answer is "technically you can do this but it's not useful in any way because you then cannot get the data or it's not usable because it's stored in a compressed format" we're not documenting it.

> 

Author: I put these two sentences on the viam-server page anyway, so I think it's fine in its current state. I will double-check about whether the storage disappears when power cycling.

{{< /expand >}}

### Supported components and services

The following components and services support data capture:

{{< tabs >}}
{{% tab name="viam-server" %}}

| Type                           |
| ------------------------------ |
| Arm                            |
| Board                          |
| Camera                         |
| Encoder                        |
| Gantry                         |
| Motor                          |
| Movement sensor (includes GPS) |
| Sensor                         |
| Servo                          |
| Vision service                 |

{{% /tab %}}
{{% tab name="viam-micro-server" %}}

<!-- prettier-ignore -->
| Type | Method |
| ---- | ------ |
| [Movement Sensor](/components/movement-sensor/) | [`AngularVelocity`](/components/movement-sensor/#getangularvelocity), [`LinearAcceleration`](/components/movement-sensor/#getlinearacceleration), [`LinearVelocity`](/components/movement-sensor/#getlinearvelocity) |
| [Sensor](/components/sensor/) | [`GetReadings`](/components/sensor/#getreadings) |

{{% /tab %}}
{{< /tabs >}}

## Cloud sync details

> ### Frequency

Reviewer: I think this should be explained in the parameter table where it's being configured. We may need two parameter tables for each part of the config.

Author: okay I'll remove it here

Author: I don't think we do--I think we just need viam-server versus viam-micro-server, which I'm doing in tabs

{{<imgproc src="/services/data/part-menu.png" resize="220x" declaredimensions=true alt="Machine menu with the options Rename, Restart part, View captured data, View setup instructions, View history, View debug configuration, and Delete machine">}}

{{% /tab %}}
{{% tab name="Whole machine" %}}

To view captured data for a {{< glossary_tooltip term_id="machine" text="machine" >}} (including data from all parts of a multi-part machine), click on the data icon next to the **Save** button on the top right of the app.

{{<imgproc src="/services/data/data-icon.png" resize="200x" declaredimensions=true alt="Data icon">}}

{{% /tab %}}
{{< /tabs >}}

{{< /expand >}}
{{< expand "Step 5: (Optional) Pause sync" >}}

You can pause cloud sync at any time by navigating to your machine's **CONFIGURE** tab and disabling **Syncing** for your [data management service](../).

If you have captured data that you do not want to sync, delete the data on the machine before resuming cloud sync.
To delete the data locally, `ssh` into your machine and delete the data in the directory where you capture data.

{{< /expand >}}

### Capture data selectively with filtering

> See the [Use filtering to collect and sync only certain images](/how-tos/image-data/#use-filtering-to-collect-and-sync-only-certain-images) guide.

Reviewer: I think this may be useful as a call out during configuration instead. The same way you've dealt with pausing sync. 
Something like "Step 6: Too much data? Limit the captured data to interesting data"

> ### Sync files from another directory

Reviewer: I think we can leave this for explanation to the attribute table and a mention in the configuration steps

Author: okay. I assumed it exited here because someone wanted to be able to link users to it, and you can't link them to steps in the expanders I don't think.

{{% /tab %}}
{{% tab name="Raw JSON example" %}}

In the example below, the data management service syncs the configured component data from `/.viam/capture` as well as all files in `/logs` every 0.1 minutes (6 seconds).

```json {class="line-numbers linkable-line-numbers"}
{
  "components": [],
  "services": [
    {
      "name": "data_manager",
      "type": "data_manager",
      "attributes": {
        "sync_interval_mins": 0.1,
        "capture_dir": "",
        "sync_disabled": false,
        "additional_sync_paths": ["/logs"]
      }
    }
  ]
}
```

{{% /tab %}}
{{< /tabs >}}

### Sync data conditionally

You can use a {{< glossary_tooltip term_id="module" text="module" >}} to sync data only when a certain logic condition is met, instead of at a regular time interval.
For example, if you rely on mobile data but have intermittent WiFi connection in certain locations or at certain times of the day, you may want to trigger sync to only occur when these conditions are met.
To set up triggers for syncing see:

{{< cards >}}
{{% card link="/how-tos/trigger-sync/" %}}
{{< /cards >}}

> ### Configure sync threads

Reviewer: Also this should go in the attribute table

If you have captured data that you do not want to sync, delete the data on the machine before resuming cloud sync.
To delete the data locally, `ssh` into your machine and delete the data in the directory where you capture data.

{{< /expand >}}

## Sync data conditionally

You can use a {{< glossary_tooltip term_id="module" text="module" >}} to sync data only when a certain logic condition is met, instead of at a regular time interval.
For example, if you rely on mobile data but have intermittent WiFi connection in certain locations or at certain times of the day, you may want to trigger sync to only occur when these conditions are met.
To set up triggers for syncing see:

{{< cards >}}
{{% card link="/how-tos/trigger-sync/" %}}
{{< /cards >}}

## API

The data management service supports the following methods:

{{< readfile "/static/include/services/apis/generated/data_manager-table.md" >}}

The data client API supports a separate set of methods that allow you to upload and export data to and from the Viam app.
For information about that API, see [Data Client API](/appendix/apis/data-client/).

{{% alert title="Tip" color="tip" %}}

> The following code examples assume that you have a machine configured with a data management service called `"my_data_service"`, and that you add the required code to connect to your machine and import any required packages at the top of your code file.

Reviewer: (We don't actually ever use that name in the samples I think)
```suggestion
The following code examples assume that you have a machine configured with a data management service, and that you add the required code to connect to your machine and import any required packages at the top of your code file.
```
